File: Candy/Candy.cpp
class Solution {
public:
    int candy(vector<int> &ratings) {
        int total = 0;
        std::vector<int> alloc;
        alloc.push_back(1);
        for(int i = 0; i < ratings.size() - 1; ++i){
            
            if(ratings[i + 1] > ratings[i]){
                int next = alloc.back() + 1;
                alloc.push_back(next);
            }
            else{
                alloc.push_back(1);
            }
            
        }
        
        for(int i = ratings.size() - 1; i > 0; --i){
            if(ratings[i] < ratings[i - 1]){
                if(alloc[i] >= alloc[i - 1]){
                    alloc[i - 1] = alloc[i] + 1;
                }
            }
            total += alloc[i];
        }
        total += alloc[0];
        
        return total;
    }
};

File: Candy/Candy.java
public class Solution {
    public int candy(int[] ratings) {
        int total = 1;
        int[] candies = new int[ratings.length];
        candies[0] = 1;
        int minimum = 1;
        
        for(int i = 1; i < ratings.length; ++i){
            if(ratings[i] > ratings[i - 1]){
                candies[i] = candies[i - 1] + 1;
                total += candies[i];
            }
            else{
                candies[i] = 1;
                ++total;
            }
        }
        
        for(int i = ratings.length - 2; i >= 0; --i){
            if(ratings[i + 1] < ratings[i]){
                int prev = candies[i];
                candies[i] = Math.max(candies[i], candies[i + 1] + 1);
                total += (candies[i] - prev);
            }
        }
        
        return total;
    }
}
